cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,364 @@
+type FirewallManager interface {
+	// Get returns the current public access firewall representation (i.e., the DO Firewall object)
+	// if it exists and nil if it does not exist.
+	Get(ctx context.Context) (*godo.Firewall, error)
+
+	// Set applies the given inbound rules to the public access firewall.
+	Set(ctx context.Context, firewallRequest *godo.FirewallRequest) error
+}
+
+// FirewallController helps to keep cloud provider service firewalls in sync.
+type FirewallController struct {
+	kubeClient         clientset.Interface
+	client             *godo.Client
+	workerFirewallTags []string
+	workerFirewallName string
+	serviceLister      corelisters.ServiceLister
+	fwManager          FirewallManager
+}
+
+// NewFirewallController returns a new firewall controller to reconcile public access firewall state.
+func NewFirewallController(
+	ctx context.Context,
+	kubeClient clientset.Interface,
+	client *godo.Client,
+	serviceInformer coreinformers.ServiceInformer,
+	fwManager FirewallManager,
+	workerFirewallTags []string,
+	workerFirewallName string,
+) *FirewallController {
+	fc := &FirewallController{
+		kubeClient:         kubeClient,
+		client:             client,
+		workerFirewallTags: workerFirewallTags,
+		workerFirewallName: workerFirewallName,
+		fwManager:          fwManager,
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
+				defer cancel()
+				err := fc.ensureReconciledFirewall(ctx)
+				if err != nil {
+					klog.Errorf("failed to reconcile worker firewall after adding a resource object: %s", err)
+				}
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
+				defer cancel()
+				err := fc.ensureReconciledFirewall(ctx)
+				if err != nil {
+					klog.Errorf("failed to reconcile worker firewall after updating a resource object: %s", err)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
+				defer cancel()
+				err := fc.ensureReconciledFirewall(ctx)
+				if err != nil {
+					klog.Errorf("failed to reconcile worker firewall after deleting a resource object: %s", err)
+				}
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}, fm *firewallManagerOp, fwReconcileFrequency time.Duration) {
+	wait.Until(func() {
+		err := fc.actualRun(stopCh, fm, fwReconcileFrequency)
+		if err != nil {
+			klog.Errorf("failed to run firewall controller loop: %v", err)
+		}
+	}, fwReconcileFrequency, stopCh)
+}
+
+func (fc *FirewallController) actualRun(stopCh <-chan struct{}, fm *firewallManagerOp, fwReconcileFrequency time.Duration) error {
```

Comment:
I think we do not need to pass in `fm` since the firewall manager is also available as a field on the `fc` struct.

Reply:
`fc.fwManager` itself does not allow for access to anything other than `Get` or `Set`, it is just an interface. Whereas `firewallManagerOp` is a concrete implementation of `fwManager` but has some added things.

```go
// firewallManagerOp manages the interaction with the DO Firewalls API.
type firewallManagerOp struct {
	client             *godo.Client
	fwCache            firewallCache
	workerFirewallName string
	workerFirewallTags []string
}
```

vs 

```go
// FirewallManager retrieves and stores firewall representations.
type FirewallManager interface {
	// Get returns the current public access firewall representation (i.e., the DO Firewall object)
	// if it exists and nil if it does not exist.
	Get(ctx context.Context) (*godo.Firewall, error)

	// Set applies the given firewall request to the public access firewall.
	Set(ctx context.Context, firewallRequest *godo.FirewallRequest) error
}
```

Maybe I'm missing something w/ the relation here @timoreimann ?